test/difference.jl

```diff
@@ -0,0 +1,57 @@
+function test_difference(ε::Real, x, dx)
+    y = x + ε * dx
+    dx_diff = difference(ε, y, x)
+    @test typeof(dx) == typeof(dx_diff)
+end
+
+@testset "difference" begin
+
+    @testset "$(typeof(x))" for (ε, x) in [
```

Comment:
These tests cover everything covered in the `to_vec` tests other than `Dict`, which doesn't currently play nicely / I'm not entirely sure how best to construct a tangent, nor how to add it to the primal.